huawei_vrp/docker/launch.py

```diff
@@ -0,0 +1,281 @@
+            cmd=f"ssh user {self.username} authentication-type password ", wait="]"
+        )
+        self.wait_write(cmd=f"ssh user {self.username} service-type all ", wait="]")
+        self.wait_write(cmd="stelnet server enable", wait="]")
+
+        # NETCONF
+        self.wait_write(cmd="snetconf server enable", wait="]")
+        self.wait_write(cmd="netconf", wait="]")
+        self.wait_write(cmd="protocol inbound ssh port 830", wait="]")
+        self.wait_write(cmd="quit", wait="]")
+
+        # Error: The system is busy in building configuration. Please wait for a moment...
+        while True:
+            self.wait_write(cmd="commit", wait=None)
+            (idx, match, res) = self.tn.expect([rb"\[~"], 1)
+            if match and idx == 0:
+                break
+            time.sleep(5)
+
+        self.wait_write(cmd="return", wait=None)
+        self.wait_write(cmd="save", wait=">")
+        self.wait_write(cmd="undo mmi-mode enable", wait=">")
+
+    def startup_config(self):
+        if not os.path.exists(STARTUP_CONFIG_FILE):
+            self.logger.trace(f"Startup config file {STARTUP_CONFIG_FILE} not found")
+            return
+
+        if self.vm_type == "CE12800":
+            with open(STARTUP_CONFIG_FILE, "r+") as file:
+                cfg = file.read()
+                modified = False
+
+                if "device board 1 " not in cfg:
+                    cfg = "device board 1 board-type CE-LPUE\n" + cfg
+                    modified = True
+
+                if "interface NULL0" not in cfg:
+                    cfg = cfg + "\ninterface NULL0"
+                    modified = True
+
+                if modified:
+                    file.seek(0)
+                    file.write(cfg)
+                    file.truncate()
+
+
+        self.wait_write(cmd="mmi-mode enable", wait=None)
+        self.wait_write(cmd="system-view", wait=None)
+        self.wait_write(
+            cmd=f"ssh user {self.username} sftp-directory cfcard:", wait="]"
+        )
+        self.wait_write(cmd="sftp server enable", wait="]")
+        self.wait_write(cmd="commit", wait="]")
+
+        time.sleep(5)
+
+        ssh_client = paramiko.SSHClient()
+        ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        ssh_client.connect(
+            hostname="localhost", username=self.username, password=self.password
+        )
+        sftp_client = ssh_client.open_sftp()
+
+        sftp_client.put(STARTUP_CONFIG_FILE, "containerlab.cfg")
+        print(f"File '{STARTUP_CONFIG_FILE}' successfully transferred")
+
+        sftp_client.close()
+        ssh_client.close()
+
+        self.wait_write(cmd="return", wait="]")
+        self.wait_write(cmd="startup saved-configuration containerlab.cfg", wait=">")
+        self.wait_write(cmd="reboot fast", wait=">")
+        self.wait_write(cmd="reboot", wait="#")
+        self.wait_write(cmd="", wait="The current login time is")
+        self.bootstrap_config()
```

Comment:
why do we need to call bootstrap config again? won't it overwrite the startup config essentially?

Reply:
Yes, it will overwrite the management interface and SSH configuration for the input user. The advantage of this approach is that it also enables the use of partial configurations. The idea was to merge the startup configuration with the bootstrap configuration, which is useful when the startup config doesn't include or has incorrect settings for SSH accessibility.